### PR TITLE
Onboarding

### PR DESCRIPTION
**Pull request by kaydee0502**
## User Onboarding

[Notion Task Link](https://www.notion.so/Onboarding-206142200fdd4ac8a9b96d40e21e7d76)

### CODE REVIEW

File: .codeclimate.yml [hunk 1]
```diff
@@ -13,3 +13,5 @@ exclude_patterns:
   - "devsnest/spec/"
   - "devsnest/db/schema.rb"
   - "devsnest/lib/tasks/youtube.rake"
+  - "devsnest/config/routes.rb"
```
Comment (vedansh) [inline .codeclimate.yml:16]: why?, isn't this an important file

File: devsnest/app/controllers/application_controller.rb [hunk 2]
```diff
@@ -62,4 +62,10 @@ def configure_permitted_parameters
   def check_username(username)
     username.match(/^(?!.*\.\.)(?!.*\.$)[^\W][\w.]{4,29}$/).nil?
   end
+
+  def admin_auth
+    return true if @current_user.user_type == 'admin'
```
Comment (vedansh) [inline devsnest/app/controllers/application_controller.rb:67]: return true if @current_user.present? and @current_user.user_type == 'admin'

File: devsnest/app/controllers/api/v1/admin_controller.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+module Api
+  module V1
+    # admin methods
+    class AdminController < ApplicationController
+      include JSONAPI::ActsAsResourceController
+      before_action :user_auth
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/admin_controller.rb:8]: user_auth may not be required

File: devsnest/app/controllers/api/v1/admin_controller.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+      before_action :admin_auth
+
+      def onboard_details
+        @onboards = User.all
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/admin_controller.rb:12]: We don't need all users, can we only send CSV data for users who filled the group form and haven't gotten a team yet, or something around last 2 weeks

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 5]
```diff
@@ -138,6 +140,18 @@ def update_username
           params['data']['attributes']['update_count'] = context[:user].update_count + 1
         end
       end
+
+      def onboard
+        ActionController::Parameters.permit_all_parameters = true
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/users_controller.rb:145]: do we need this line?

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 6]
```diff
@@ -138,6 +140,18 @@ def update_username
...
+        return render_error('Form already filled') if @current_user.is_form_filled
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/users_controller.rb:146]: can we rename this variable to specifically point to discord form

File: devsnest/app/controllers/api/v1/admin_controller.rb (hunk 4 above)
Comment (vedansh) [reply devsnest/app/controllers/api/v1/admin_controller.rb:12]: I saw that you have applied the filter in the to_csv function, that will result in multiple db calls, add these filters here

Comment (kaydee0502) [reply devsnest/app/controllers/api/v1/admin_controller.rb]: done

File: devsnest/app/controllers/api/v1/admin_controller.rb (hunk 3 above)
Comment (kaydee0502) [reply devsnest/app/controllers/api/v1/admin_controller.rb]: removed, just using admin_auth

File: devsnest/app/controllers/application_controller.rb (hunk 2 above)
Comment (kaydee0502) [reply devsnest/app/controllers/application_controller.rb]: done

File: .codeclimate.yml (hunk 1 above)
Comment (kaydee0502) [reply .codeclimate.yml]: removed

File: devsnest/app/controllers/api/v1/users_controller.rb (hunk 6 above)
Comment (kaydee0502) [reply devsnest/app/controllers/api/v1/users_controller.rb]: changed to `is_discord_form_filled`

File: devsnest/app/controllers/api/v1/admin_controller.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+      include JSONAPI::ActsAsResourceController
+      before_action :admin_auth
+
+      def onboard_details
+        @onboards = User.where('created_at >= :span and is_discord_form_filled', span: 2.weeks.ago, is_discord_form_filled: true)
+        response.headers['Content-Type'] = 'text/csv; charset=UTF-8; header=present'
+        response.headers['Content-Disposition'] = 'attachment; filename=onboard_result.csv'
+        render plain: @onboards.to_csv
+      end
+    end
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/admin_controller.rb:16]: going to this url, directly downloads the csv?

Comment (kaydee0502) [reply devsnest/app/controllers/api/v1/admin_controller.rb:16]: yes

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 8]
```diff
@@ -138,6 +138,18 @@ def update_username
           params['data']['attributes']['update_count'] = context[:user].update_count + 1
         end
       end
+
+      def onboard
+        updatable_params = params.require(:data).permit(attributes: [:discord_username, :discord_id, :name, :work_exp, :known_from, :dsa_skill, :webd_skill, :is_discord_form_filled, :college_id])
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/users_controller.rb:143]: awesome

**Comment by vedansh** [comment]
@kaydee0502 Looks good, just add some more specs

File: devsnest/spec/models/user_spec.rb [hunk 9]
```diff
@@ -45,4 +45,10 @@
       expect(User.find_by(google_id: 9000)).to be_present
     end
   end
+
+  context 'check to_csv method' do
+    it 'checks for empty csv' do
+      expect(described_class.to_csv).to eq "id,discord_username,discord_id,name,grad_year,school,work_exp,known_from,dsa_skill,webd_skill\n"
+    end
+  end
```
Comment (vedansh) [inline devsnest/spec/models/user_spec.rb:53]: 1. also test the controller
2. Add a test case when there is data

File: devsnest/spec/models/user_spec.rb [hunk 10]
```diff
@@ -45,4 +45,17 @@
       expect(User.find_by(google_id: 9000)).to be_present
     end
   end
+
+  context 'check to_csv method' do
+    it 'checks for empty csv' do
+      expect(described_class.to_csv).to eq "id,discord_username,discord_id,name,grad_year,school,work_exp,known_from,dsa_skill,webd_skill\n"
+    end
+
+    # it 'checks for some fields' do
```
Comment (vedansh) [inline devsnest/spec/models/user_spec.rb:54]: why is this commented?

File: devsnest/spec/support/jsonapi_matchers.rb [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+  end
+  
+  RSpec::Matchers.define :have_attributes do |*attrs|
+    match do |actual|
+      return false unless actual.key?(:attributes)
+  
+      attrs.all? { |attr| actual[:attributes].key?(attr) }
+    end
+  end
+
```
Comment (vedansh) [inline devsnest/spec/support/jsonapi_matchers.rb:31]: this file has wrong indentation

**Review by sam0000-cmd — APPROVED** [review]
Noice

**Pull request merged**